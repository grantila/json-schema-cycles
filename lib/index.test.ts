import * as path from "path";

import {
	analyzeTypes,
	analyzeTypesFast,
	sortTypeAnalysisFullResult,
	sortTypeAnalysisFastResult,
} from './index'


const fixturePath = path.resolve( __dirname, '..', 'fixtures' );
const largeJson = require( `${fixturePath}/large-json-schema.json` );


function expectEqual< T >( got: T, expected: T )
{
	expect( got ).toStrictEqual( expected );
}

describe( "analyzeTypes", ( ) =>
{
	it( "no definition", ( ) =>
	{
		expectEqual(
			analyzeTypes( { } ),
			{
				entrypoints: [ ],
				cycles: [ ],
				all: [ ],
				dependencies: [ ],
				dependents: [ ],
				graph: [ ],
			}
		);
	} );

	it( "no types", ( ) =>
	{
		expectEqual(
			analyzeTypes( { definitions: { } } ),
			{
				entrypoints: [ ],
				cycles: [ ],
				all: [ ],
				dependencies: [ ],
				dependents: [ ],
				graph: [ ],
			}
		);
	} );

	it( "simple type", ( ) =>
	{
		expectEqual(
			analyzeTypes( { definitions: { User: { } } } ),
			{
				entrypoints: [ ],
				cycles: [ ],
				all: [ ],
				dependencies: [ ],
				dependents: [ ],
				graph: [ [ 'User', [ ] ] ],
			}
		);
	} );

	it( "single recursive type", ( ) =>
	{
		const definitions = {
			User: {
				type: 'object',
				properties: {
					parent: { $ref: '#/definitions/User' },
				},
			},
		};
		expectEqual(
			analyzeTypes( { definitions } ),
			{
				entrypoints: [ ],
				cycles: [ [ 'User' ] ],
				all: [ 'User' ],
				dependencies: [ ],
				dependents: [ ],
				graph: [ [ 'User', [ 'User' ] ] ],
			}
		);
	} );

	it( "complex typings", ( ) =>
	{
		const jsonSchema = {
			definitions: {
				Link: {}, // Non-cyclic but dependency of Message
				Subscriber: {
					type: 'object',
					properties: {
						user: { $ref: '#/definitions/User' },
					},
				},
				Message: {
					type: 'object',
					properties: {
						replyTo: { $ref: '#/definitions/Message' },
						link: { $ref: '#/definitions/Link' },
						subscriber: { $ref: '#/definitions/Subscriber' },
					},
				},
				User: {
					type: 'object',
					properties: {
						parent: { $ref: '#/definitions/User' },
						lastMessage: { $ref: '#/definitions/Message' },
					},
				},
				DM: {
					type: 'object',
					properties: {
						lastUser: { $ref: '#/definitions/User' },
					},
				},
				Actions: {
					type: 'object',
					properties: {
						dms: {
							type: 'array',
							items: { $ref: '#/definitions/DM' },
						},
					},
				},
				// Has dependencies, but nothing cyclic
				Product: { $ref: '#/non-local-dep/Foo' },
				Cart: {
					type: 'array',
					items: { $ref: '#/definitions/Product' },
				},
			}
		};

		expectEqual(
			sortTypeAnalysisFullResult( analyzeTypes( jsonSchema ) ),
			sortTypeAnalysisFullResult( {
				entrypoints: [
					[ "DM" ],
					[ "Actions", "DM" ],
				],
				cycles: [
					[ 'User' ],
					[ 'Message' ],
					[ 'User', 'Message', 'Subscriber' ],
				],
				all: [ 'User', 'Message', 'DM', 'Actions', 'Subscriber' ],
				dependencies: [ "Link" ],
				dependents: [ ],
				graph: [
					[ 'Link', [ ] ],
					[ 'Subscriber', [ 'User' ] ],
					[ 'Message', [ 'Message', 'Link', 'Subscriber' ] ],
					[ 'User', [ 'Message', 'User' ] ],
					[ 'DM', [ 'User' ] ],
					[ 'Actions', [ 'DM' ] ],
					[ 'Product', [ ] ],
					[ 'Cart', [ 'Product' ] ],
				],
			} ),
		);
	} );
} );

describe( "analyzeTypesFast", ( ) =>
{
	it( "no definition", ( ) =>
	{
		expectEqual(
			analyzeTypesFast( { } ),
			{
				cyclic: [ ],
				dependencies: [ ],
				dependents: [ ],
				graph: [ ],
			}
		);
	} );

	it( "no types", ( ) =>
	{
		expectEqual(
			analyzeTypesFast( { definitions: { } } ),
			{
				cyclic: [ ],
				dependencies: [ ],
				dependents: [ ],
				graph: [ ],
			}
		);
	} );

	it( "simple type", ( ) =>
	{
		expectEqual(
			analyzeTypesFast( { definitions: { User: { } } } ),
			{
				cyclic: [ ],
				dependencies: [ ],
				dependents: [ ],
				graph: [ [ 'User', [ ] ] ],
			}
		);
	} );

	it( "single recursive type", ( ) =>
	{
		const definitions = {
			User: {
				type: 'object',
				properties: {
					parent: { $ref: '#/definitions/User' },
				},
			},
		};
		expectEqual(
			analyzeTypesFast( { definitions } ),
			{
				cyclic: [ 'User' ],
				dependencies: [ ],
				dependents: [ ],
				graph: [ [ 'User', [ 'User' ] ] ],
			}
		);
	} );

	it( "complex typings", ( ) =>
	{
		const jsonSchema = {
			definitions: {
				Link: {}, // Non-cyclic but dependency of Message
				Subscriber: {
					type: 'object',
					properties: {
						user: { $ref: '#/definitions/User' },
					},
				},
				Message: {
					type: 'object',
					properties: {
						replyTo: { $ref: '#/definitions/Message' },
						link: { $ref: '#/definitions/Link' },
						subscriber: { $ref: '#/definitions/Subscriber' },
					},
				},
				User: {
					type: 'object',
					properties: {
						parent: { $ref: '#/definitions/User' },
						lastMessage: { $ref: '#/definitions/Message' },
					},
				},
				DM: {
					type: 'object',
					properties: {
						lastUser: { $ref: '#/definitions/User' },
					},
				},
				Actions: {
					type: 'object',
					properties: {
						dms: {
							type: 'array',
							items: { $ref: '#/definitions/DM' },
						},
					},
				},
				// Has dependencies, but nothing cyclic
				Product: { $ref: '#/non-local-dep/Foo' },
				Cart: {
					type: 'array',
					items: { $ref: '#/definitions/Product' },
				},
			}
		};

		expectEqual(
			sortTypeAnalysisFastResult( analyzeTypesFast( jsonSchema ) ),
			sortTypeAnalysisFastResult( {
				cyclic: [ 'User', 'Message', 'DM', 'Actions', 'Subscriber' ],
				dependencies: [ "Link" ],
				dependents: [ ],
				graph: [
					[ 'Link', [ ] ],
					[ 'Subscriber', [ 'User' ] ],
					[ 'Message', [ 'Message', 'Link', 'Subscriber' ] ],
					[ 'User', [ 'Message', 'User' ] ],
					[ 'DM', [ 'User' ] ],
					[ 'Actions', [ 'DM' ] ],
					[ 'Product', [ ] ],
					[ 'Cart', [ 'Product' ] ],
				],
			} ),
		);
	} );
} );

describe( "sortTypeAnalysisFullResult", ( ) =>
{
	it( "should sort properly", ( ) =>
	{
		const sorted = sortTypeAnalysisFullResult( {
			all: [ ],
			cycles: [ ],
			dependencies: [ 'b', 'a' ],
			dependents: [ ],
			entrypoints: [ ],
			graph: [
				[ 'x', [ 'a', 'c', 'b' ] ],
				[ 'f', [ 'g', 'h' ] ],
				[ 'g', [ 'h', 'f', 'g' ] ],
			],
		} );

		expect( sorted ).toStrictEqual( {
			all: [ ],
			cycles: [ ],
			dependencies: [ 'a', 'b' ],
			dependents: [ ],
			entrypoints: [ ],
			graph: [
				[ 'f', [ 'g', 'h' ] ],
				[ 'g', [ 'f', 'g', 'h' ] ],
				[ 'x', [ 'a', 'b', 'c' ] ],
			],
		} );
	} );
} );

describe( "sortTypeAnalysisFastResult", ( ) =>
{
	it( "should sort properly", ( ) =>
	{
		const sorted = sortTypeAnalysisFastResult( {
			cyclic: [ 'b', 'a', 'c' ],
			dependencies: [ 'y', 'x' ],
			dependents: [ 'z', 'w' ],
			graph: [
				[ 'x', [ 'a', 'c', 'b' ] ],
				[ 'f', [ 'g', 'h' ] ],
				[ 'g', [ 'h', 'f', 'g' ] ],
			],
		} );

		expect( sorted ).toStrictEqual( {
			cyclic: [ 'a', 'b', 'c' ],
			dependencies: [ 'x', 'y' ],
			dependents: [ 'w', 'z' ],
			graph: [
				[ 'f', [ 'g', 'h' ] ],
				[ 'g', [ 'f', 'g', 'h' ] ],
				[ 'x', [ 'a', 'b', 'c' ] ],
			],
		} );
	} );
} );

describe( "large input (fast mode)", ( ) =>
{
	it( "should handle really large json schema", ( ) =>
	{
		const { graph, ...analysis } = analyzeTypesFast( largeJson );

		expect( analysis ).toMatchSnapshot( );
	} );
} );
