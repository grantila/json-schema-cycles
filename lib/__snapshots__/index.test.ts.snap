// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`large input (fast mode) should handle really large json schema 1`] = `
{
  "cyclic": [
    "Node",
    "ArrayExpression",
    "Expression",
    "AssignmentExpression",
    "LVal",
    "Identifier",
    "Decorator",
    "TypeAnnotation",
    "FlowType",
    "ArrayTypeAnnotation",
    "FunctionTypeAnnotation",
    "TypeParameterDeclaration",
    "TypeParameter",
    "FunctionTypeParam",
    "GenericTypeAnnotation",
    "QualifiedTypeIdentifier",
    "TypeParameterInstantiation",
    "InterfaceTypeAnnotation",
    "InterfaceExtends",
    "ObjectTypeAnnotation",
    "ObjectTypeProperty",
    "ObjectTypeSpreadProperty",
    "ObjectTypeIndexer",
    "ObjectTypeCallProperty",
    "ObjectTypeInternalSlot",
    "IntersectionTypeAnnotation",
    "NullableTypeAnnotation",
    "TupleTypeAnnotation",
    "TypeofTypeAnnotation",
    "UnionTypeAnnotation",
    "TSTypeAnnotation",
    "TSType",
    "TSFunctionType",
    "TSTypeParameterDeclaration",
    "TSTypeParameter",
    "RestElement",
    "TSConstructorType",
    "TSTypeReference",
    "TSEntityName",
    "TSQualifiedName",
    "TSTypeParameterInstantiation",
    "TSTypePredicate",
    "TSTypeQuery",
    "TSImportType",
    "TSTypeLiteral",
    "TSTypeElement",
    "TSCallSignatureDeclaration",
    "TSConstructSignatureDeclaration",
    "TSPropertySignature",
    "TSMethodSignature",
    "TSIndexSignature",
    "TSArrayType",
    "TSTupleType",
    "TSNamedTupleMember",
    "TSOptionalType",
    "TSRestType",
    "TSUnionType",
    "TSIntersectionType",
    "TSConditionalType",
    "TSInferType",
    "TSParenthesizedType",
    "TSTypeOperator",
    "TSIndexedAccessType",
    "TSMappedType",
    "TSExpressionWithTypeArguments",
    "MemberExpression",
    "PrivateName",
    "AssignmentPattern",
    "ObjectPattern",
    "ObjectProperty",
    "PatternLike",
    "ArrayPattern",
    "TSParameterProperty",
    "BinaryExpression",
    "CallExpression",
    "SpreadElement",
    "ConditionalExpression",
    "FunctionExpression",
    "Pattern",
    "BlockStatement",
    "Statement",
    "BreakStatement",
    "ContinueStatement",
    "DoWhileStatement",
    "ExpressionStatement",
    "ForInStatement",
    "VariableDeclaration",
    "VariableDeclarator",
    "ForStatement",
    "FunctionDeclaration",
    "IfStatement",
    "LabeledStatement",
    "ReturnStatement",
    "SwitchStatement",
    "SwitchCase",
    "ThrowStatement",
    "TryStatement",
    "CatchClause",
    "WhileStatement",
    "WithStatement",
    "ClassDeclaration",
    "ClassBody",
    "ClassMethod",
    "ClassPrivateMethod",
    "ClassProperty",
    "ClassPrivateProperty",
    "TSDeclareMethod",
    "ClassImplements",
    "ExportAllDeclaration",
    "ImportAttribute",
    "ExportDefaultDeclaration",
    "TSDeclareFunction",
    "ExportNamedDeclaration",
    "Declaration",
    "ImportDeclaration",
    "ImportSpecifier",
    "ImportDefaultSpecifier",
    "ImportNamespaceSpecifier",
    "DeclareClass",
    "DeclareFunction",
    "DeclaredPredicate",
    "Flow",
    "DeclareInterface",
    "DeclareModule",
    "DeclareModuleExports",
    "DeclareTypeAlias",
    "DeclareOpaqueType",
    "DeclareVariable",
    "DeclareExportDeclaration",
    "ExportSpecifier",
    "ExportNamespaceSpecifier",
    "InterfaceDeclaration",
    "OpaqueType",
    "TypeAlias",
    "TypeCastExpression",
    "EnumDeclaration",
    "EnumBooleanBody",
    "EnumBooleanMember",
    "EnumNumberBody",
    "EnumNumberMember",
    "EnumStringBody",
    "EnumStringMember",
    "EnumDefaultedMember",
    "EnumSymbolBody",
    "TSInterfaceDeclaration",
    "TSInterfaceBody",
    "TSTypeAliasDeclaration",
    "TSEnumDeclaration",
    "TSEnumMember",
    "TSModuleDeclaration",
    "TSModuleBlock",
    "ExportDefaultSpecifier",
    "ForOfStatement",
    "TSImportEqualsDeclaration",
    "TSExportAssignment",
    "TSNamespaceExportDeclaration",
    "LogicalExpression",
    "NewExpression",
    "ObjectExpression",
    "ObjectMethod",
    "SequenceExpression",
    "ParenthesizedExpression",
    "UnaryExpression",
    "UpdateExpression",
    "ArrowFunctionExpression",
    "ClassExpression",
    "MetaProperty",
    "TaggedTemplateExpression",
    "TemplateLiteral",
    "YieldExpression",
    "AwaitExpression",
    "OptionalMemberExpression",
    "OptionalCallExpression",
    "JSXElement",
    "JSXOpeningElement",
    "JSXMemberExpression",
    "JSXAttribute",
    "JSXFragment",
    "JSXExpressionContainer",
    "JSXSpreadChild",
    "JSXSpreadAttribute",
    "JSXClosingElement",
    "BindExpression",
    "DoExpression",
    "RecordExpression",
    "TupleExpression",
    "TSAsExpression",
    "TSTypeAssertion",
    "TSNonNullExpression",
    "Binary",
    "Block",
    "Program",
    "BlockParent",
    "StaticBlock",
    "Class",
    "CompletionStatement",
    "Conditional",
    "EnumBody",
    "EnumMember",
    "ExportDeclaration",
    "ExpressionWrapper",
    "File",
    "FlowDeclaration",
    "FlowPredicate",
    "For",
    "ForXStatement",
    "Function",
    "FunctionParent",
    "Immutable",
    "JSX",
    "Literal",
    "Loop",
    "Method",
    "ModuleDeclaration",
    "ModuleSpecifier",
    "ObjectMember",
    "PipelineBareFunction",
    "PipelineTopicExpression",
    "Placeholder",
    "Private",
    "Property",
    "Pureish",
    "RestProperty",
    "Scopable",
    "SpreadProperty",
    "Terminatorless",
    "UnaryLike",
    "UserWhitespacable",
    "While",
    "Aliases",
  ],
  "dependencies": [
    "AnyTypeAnnotation",
    "ArgumentPlaceholder",
    "BooleanTypeAnnotation",
    "BooleanLiteralTypeAnnotation",
    "NullLiteralTypeAnnotation",
    "ExistsTypeAnnotation",
    "Variance",
    "StringLiteral",
    "MixedTypeAnnotation",
    "EmptyTypeAnnotation",
    "NumberLiteralTypeAnnotation",
    "NumberTypeAnnotation",
    "StringLiteralTypeAnnotation",
    "StringTypeAnnotation",
    "SymbolTypeAnnotation",
    "ThisTypeAnnotation",
    "VoidTypeAnnotation",
    "TSAnyKeyword",
    "TSBooleanKeyword",
    "TSBigIntKeyword",
    "TSIntrinsicKeyword",
    "TSNeverKeyword",
    "TSNullKeyword",
    "TSNumberKeyword",
    "TSObjectKeyword",
    "TSStringKeyword",
    "TSSymbolKeyword",
    "TSUndefinedKeyword",
    "TSUnknownKeyword",
    "TSVoidKeyword",
    "TSThisType",
    "Noop",
    "NumericLiteral",
    "TSLiteralType",
    "BooleanLiteral",
    "BigIntLiteral",
    "V8IntrinsicIdentifier",
    "JSXIdentifier",
    "JSXNamespacedName",
    "DebuggerStatement",
    "EmptyStatement",
    "DeclareExportAllDeclaration",
    "InferredPredicate",
    "TSExternalModuleReference",
    "DirectiveLiteral",
    "Directive",
    "NullLiteral",
    "RegExpLiteral",
    "ThisExpression",
    "Super",
    "TemplateElement",
    "Import",
    "JSXOpeningFragment",
    "JSXClosingFragment",
    "JSXText",
    "JSXEmptyExpression",
    "PipelinePrimaryTopicReference",
    "DecimalLiteral",
    "InterpreterDirective",
    "CommentBlock",
    "CommentLine",
    "FlowBaseAnnotation",
    "NumberLiteral",
    "RegexLiteral",
    "TSBaseType",
  ],
  "dependents": [
    "Comment",
  ],
}
`;
